api/service/legacysync/helpers.go

```diff
@@ -58,39 +58,39 @@ func createSyncConfig(syncConfig *SyncConfig, peers []p2p.Peer, shardID uint32)
 	}
 	// limit the number of dns peers to connect
 	randSeed := time.Now().UnixNano()
-	peers = limitNumPeers(peers, randSeed)
+	targetSize, peers := limitNumPeers(peers, randSeed)
 
 	utils.Logger().Debug().
-		Int("len", len(peers)).
+		Int("peers count", len(peers)).
+		Int("target size", targetSize).
 		Uint32("shardID", shardID).
 		Msg("[SYNC] CreateSyncConfig: len of peers")
 
-	if len(peers) == 0 {
+	if targetSize == 0 {
```

Comment:
Why this change?

Reply:
Basically, what we were doing in this function was getting a list of peers, then randomly choosing a set of peers from that list, and then connecting to the peers in the selected set. If the connection to any peer in that set is broken and the client has failed to build, we remove that peer from the list. So, imagine the set contains 3 peers. If two of them are broken, it leaves only one peer connected to the node. What I did was change a bit of this flow. So, we first shuffle the peers, but we still have all the peers in our set. Then we try to connect to N peers, where N is targetSize. After we connect to those N peers, we remove the rest from the peer set. By this way we can always be sure that the peers list has maximum online nodes and we add maximum available random online peers to list. targetSize is a random number between low and high bounds, and the final number of peers will be set by that number. This part is the same as before. But we remove extra peers from the list later after initializing connections; that's why we use this variable. 